src/console/man/om_points.1
.\" Text automatically generated by txt2man
.TH om_points  "02 April 2013" "" ""
.SH NAME
\fBom_points \fP- retrieve locality data from a specified source
\fB
.SH SYNOPSIS
.nf
.fam C
     \fBom_points\fP [-] \fIv\fP \fB--version\fP | \fIl\fP \fB--list\fP | \fIs\fP \fB--source\fP \fIPATH/URL\fP \fIn\fP \fB--name\fP \fIPOINT_LABEL\fP [ \fIw\fP \fB--wkt\fP \fISPATIAL_REF_WKT\fP ] [ \fIo\fP \fB--type\fP \fIOUTPUT_TYPE\fP ] [ \fB--split\fP \fIPROPORTION\fP \fB--file1\fP \fIFILE\fP \fB--file2\fP \fIFILE\fP ] [ \fB--log-level\fP \fILEVEL\fP ]

.fam T
.fi
.fam T
.fi
.SH DESCRIPTION
\fBom_points\fP is a command line tool to retrieve locality data (occurrence points) from a specified source through one of the available drivers in openModeller. Locality data is one of the required inputs to create distribution models.
.PP
The following drivers are currently implemented in openModeller:
.PP
TXT: Delimited text file. This driver can read points from a TAB or comma separated text file located in the file system. The following columns must be present: point id, point label, longitude, latitude, abundance (this one is optional). This driver is always be available, and the same format can be used to display points retrieved in another format.
.PP
XML: openModeller serialized XML for occurrences. This driver is always be available, and the same format can be used to display points retrieved in another format. The XML produced can be used to specify points in an openModeller sampler. Two tags can be returned: <presence> and/or <absence>. Both should follow the OccurrencesType defined in http://openmodeller.cria.org.br/xml/1.0/openModeller.xsd
.PP
GBIF: GBIF Web Service for occurrence points. This driver is only available if openModeller was compiled with libcurl. Currently, this is the access point that should be specified in the source: http://data.gbif.org/ws/rest/occurrence/list
.PP
TerraLib: Points stored in a TerraLib database. This driver is only available if openModeller was compiled with TerraLib. Sources must follow the pattern: terralib>dbUsername>dbPassword@dbType>dbHost>dbName>dbPort>layerName>tableName>columnName. Where dbType can be PostgreSQL (default port 5432).
.PP
TAPIR: TAPIR Web Service using DarwinCore 1.4 and the geospatial extension. This driver is only available if openModeller was compiled with libcurl. The source must be a TAPIR access point.
.SH OPTIONS
.TP
.B
-\fIv\fP, \fB--version\fP
Display version info.
.TP
.B
-\fIl\fP, \fB--list\fP
List all available drivers to load point data (driver id: description (input/output)).
.TP
.B
-\fIs\fP, \fB--source\fP
Source of locality data. It can be a path in the file system (for TXT and XML drivers) or a URL (for GBIF or TAPIR drivers) or a special string (for the TerraLib driver).
.TP
.B
-\fIn\fP, \fB--name\fP
Name (label) to filter points. Each point in a source has a label. You must specify a label to be used for filtering the desired points.
.TP
.B
-\fIw\fP, \fB--wkt\fP
Spatial reference of the points. Optional. Defaults to lat/long datum WGS84. This is only useful if the output will be in XML, when the coordinateSystem tag is displayed. Please note that most drivers (GBIF, TAPIR and TerraLib) already assume that the points are in lat/long datum WGS84. When specified, the value must be in Well-Known Text format, with all double quotes escaped.
.TP
.B
-\fIo\fP, \fB--type\fP
Output type (can be TXT or XML). Optional. Defaults to TXT.
.TP
.B
\fB--split\fP
Split points into two subsets using the given proportion (value between 0 and 1). This can be used to create test/training partitions. Results are stored in the files specified with parameters file1 and file2. Optional.
.TP
.B
\fB--file1\fP
File name to store the first subset when points are splitted. Mandatory only when the split parameter is present.
.TP
.B
\fB--file2\fP
File name to store the second subset when points are splitted. Mandatory only when the split parameter is present.
.TP
.B
\fB--log-level\fP
openModeller log level: debug, warn, info or error. Defaults to "info".
.SH AUTHORS
Renato De Giovanni <renato at cria dot org dot br>


ChangeLog
Release 1.4.0 (?-?-?)
-------------
SVN revision: ?

	* New CONSENSUS algorithm that can be used to generate models with different algorithms by converting each individual model into a binary model given a threshold and then "merging" them into a final combined model according to a specified level of agreement.
	* Upgraded gSOAP version being used by OMWS to 2.8.13.
	* Included library API versioning, starting with 0 from this release.
	* Moved some of the compiler definitions from CMake to internal #defines, so that client programs can include openModeller headers without worrying about adding specific definitions if they want to compile directly without using CMake.
	* New command-line tool om_evaluate to return raw model values given a set of points and an environmental scenario.
	* Fixed segfault when needNormalization was called on a deserialized SVM algorithm.
	* Modified om_pseudo command-line tool to also accept XML input (in this case also producing XML output), optionally storing it into a file. When a model is specified as a parameter, presences are generated inside the model, while absences are generated outside.

Release 1.3.0 (2012-07-23)
-------------
SVN revision: 5592

	* New version of OM-MAXENT compatible with Maxent 3.3.3e.
	* Changed the deserialization code to accept XML with namespace prefixes (they are now ignored if present).
	* Fixed bug in ANN algorithm which was limiting the number of input points to 10k (seg faulting otherwise).
	* New documentation about the web service interface (OMWS), aka SOAP interface.
	* New documentation about oM Server (server-side implementation of OMWS).
	* Fixed bug in oM Server getProgress operation which was sometimes returning 100% when the job was still unfinished.
	* New oM Server configuration/script allowing automatic synchronization of layers.
	* Several improvements in the OMWS Perl command-line client.
	* Several adjustments in the SWIG Python interface, exposing more methods and objects.
	* Fixed memory leak in the Environmental Distance algorithm when the Mahalanobis metric was used after model deserialization.
	* New WCS proxy raster driver that allows WCS rasters to be identified through strings in the format: wcs>wcs_service_url>coverage_id

Release 1.2.0 (2011-07-21)
--------------------------
SVN revision: 5330

	* Changed the ENFA algorithm to force marginality to be the first factor.  Changed broken stick distribution to work on specialisation factors only.  Add new parameter to enable log display of variable loadings for retained factors.
	* om_console now accepts a second parameter (after the request file) to specify the log level.
	* Changed the ENFA algorithm allowing absence data to be treated as background points, and log transforming the geomean calculation to avoid the occasional float overflow with large numbers of presences. Fixed a bug in the RetainVariation parameter (range changed from 50-100 to 0.5-1.0). Also fixed a deserialization error.
	* Compilation now requires CMake 2.6.4.
	* New algorithm Random Forests.
	* Web service interface now accepts a new optional parameter specifying the prediction threshold to be used when calculating area statistics during projection.
	* Renamed envelope element in the getProjectionMetadata response of the web service API to distinguish it from the subelement.
	* Added control to a maximum number of concurrent jobs on the modelling server.
	* Environmental values can now be loaded from the Sampler XML when provided.
	* Web service interface now accepts optional parameters: IgnoreAbsences for the confusion matrix and UseAbsencesAsBackground for the ROC curve calculations.
	* Added support to gzip compression in the Web service interface.
	* Changed the Environmental Distance algorithm to return chi-square distribution probabilities when the metric is Mahalanobis and the maximum distance is 1.
	* Changed the Maxent algorithm allowing absence data to be treated as background points. Also fixed memory deallocation issues.

Release 1.1.0 (2010-02-05)
--------------------------
SVN revision: 5114

	* Bugfix: Interquartile and envelope size adjustments were using wrong reference values for salinity and temperature in the Aquamaps algorithm.
	* Bugfix: All GARP algorithms were crashing when the abundance value associated with one of the occurrence points had more than two digits in the integer part.
	* Increased the default number of neurons in the hidden layer of the ANN algorithm.
	* New argument "s" in om_niche allowing model outputs to be scaled.
	* Re-implemented and re-enabled Maximum entropy algorithm based on the well-known Maxent.
	* Changed environmental distance algorithm forcing the Mahalanobis distance to use an identity matrix as the covariance matrix when there's only one input point.
	* Fixed small logical error in the CSM algorithm that could affect models with a large number of layers.
	* New algorithm (Niche Mosaic) implemented by Dr. Missae Yamamoto.
	* New algorithm (ENFA) kindly implemented by Dr. Chris Yesson.

Release 1.0 (2009-05-22)
------------------------
SVN revision: 5035

	* Changed Artificial Neural Networks algorithm to avoid exceeding a maximum number of iterations when training by minimum error (parameter Epoch is used to specify the maximum number of iterations).
	* Added new method in the ConfusionMatrix class to use the lowest probability among the training presences as the threshold.
	* Changed the modelling protocol to allow createModel and testModel requests to specify what statistics should be calculated.
	* Removed model statistics parameters from om_model since they can be specified as part of the XML request.
	* All command line programs now return the same openModeller version number.
	* Changed om_niche parameters (now using getopts) and created a man page for it.
	* Implemented custom log file in the modelling server.

src/console/om_points.cpp
#include <openmodeller/om.hh>
#include <openmodeller/om_defs.hh>
#include <openmodeller/Log.hh>
#include <openmodeller/os_specific.hh>

#include "getopts/getopts.h"

#include "om_cmd_utils.hh"

#include <iostream>  // I/O 

#include <string>
#include <vector>

#include <stdexcept>

using namespace std;

void writeOutput( ostream & stream, std::string format, OccurrencesPtr presences, OccurrencesPtr absences, std::string label );

int main( int argc, char **argv ) {

  Options opts;
  int option;

  // command-line parameters (short name, long name, description, take args)
  opts.addOption( "" , "log-level", "Set the log level (debug, warn, info, error)", true );
  opts.addOption( "v", "version"  , "Display version info"                        , false );
  opts.addOption( "l", "list"     , "List available formats"                      , false );
  opts.addOption( "s", "source"   , "Source where points are located"             , true );
  opts.addOption( "n", "name"     , "Name (label) to filter points"               , true );
  opts.addOption( "w", "wkt"      , "Spatial reference in WKT"                    , true );
  opts.addOption( "o", "type"     , "Output type"                                 , true );
  opts.addOption( "" , "split"    , "Split points using the specified proportion (0,1)"  , true );
  opts.addOption( "" , "file1"    , "File name to store 1st subset (used w/ param split)", true );
  opts.addOption( "" , "file2"    , "File name to store 2nd subset (used w/ param split)", true );

  std::string log_level("info");
  bool        list_formats = false;
  std::string source("");
  std::string label("");
  std::string wkt("GEOGCS[\"WGS84\",DATUM[\"WGS84\",SPHEROID[\"WGS84\",6378137.0,298.257223563]],PRIMEM[\"Greenwich\",0.0],UNIT[\"degree\",0.017453292519943295],AXIS[\"Longitude\",EAST],AXIS[\"Latitude\",NORTH]]");
  std::string format("TXT");
  std::string split_prop_string;
  std::string file1;
  std::string file2;

  if ( ! opts.parse( argc, argv ) ) {

    opts.showHelp( argv[0] ); 
    exit(0);
  }

  // Set up any related external resources
  setupExternalResources();

  OpenModeller om;

  while ( ( option = opts.cycle() ) >= 0 ) {

    switch ( option ) {

      case 0:
        log_level = opts.getArgs( option );
        break;
      case 1:
        printf( "om_points %s\n", om.getVersion().c_str() );
        printf("This is free software; see the source for copying conditions. There is NO\n");
        printf("warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n");
        exit(0);
        break;
      case 2:
        list_formats = true;
        break;
      case 3:
        source = opts.getArgs( option );
        break;
      case 4:
        label = opts.getArgs( option );
        break;
      case 5:
        wkt = opts.getArgs( option );
        break;
      case 6:
        format = opts.getArgs( option );
        break;
      case 7:
        split_prop_string = opts.getArgs( option );
        break;
      case 8:
        file1 = opts.getArgs( option );
        break;
      case 9:
        file2 = opts.getArgs( option );
        break;
      default:
        break;
    }
  }

  // Log stuff

  Log::Level level_code = getLogLevel( log_level );

  Log::instance()->setLevel( level_code );

  // Real work

  try {

    if ( list_formats ) {

      // Process list drivers request

      std::vector<std::string> driver_ids = OccurrencesFactory::instance().getRegisteredDrivers();

      printf( "Available formats (id: description (input/output))\n" );

      for ( size_t i = 0; i < driver_ids.size(); ++i ) {

        std::string description("?");
        std::string io("I");

        if ( driver_ids[i] == "GBIF" ) {

          description = "GBIF Web Service for occurrence points";
        }
        else if ( driver_ids[i] == "TAPIR" ) {

          description = "TAPIR Web Service using DarwinCore 1.4 and the geospatial extension";
        }
        else if ( driver_ids[i] == "TerraLib" ) {

          description = "Points stored in a TerraLib database";
        }
        else if ( driver_ids[i] == "TXT" ) {

          description = "Delimited text file (id, label, long, lat, abundance)";

          io = "I/O";
        }
        else if ( driver_ids[i] == "XML" ) {

          description = "openModeller serialized XML for occurrences";

          io = "I/O";
        }

        printf( "  %s: %s (%s)\n", driver_ids[i].c_str(), description.c_str(), io.c_str() );
      }

      return 0;
    }


    // Check requirements
    if ( source.empty() ) {

      printf( "Please specify a source to load points from.\n");
      exit(-1);
    }
    if ( label.empty() ) {

      printf( "Please specify a name to filter points.\n");
      exit(-1);
    }

    // Check parameter split
    double split_prop = 0.0;

    if ( ! split_prop_string.empty() ) {

      if ( file1.empty() || file2.empty() ) {

        printf( "When splitting points, you need to specify file1 and file2.\n");
        exit(-1);
      }

      split_prop = atof( split_prop_string.c_str() );

      if ( split_prop <= 0.0 || split_prop >= 1.0 ) {

        printf( "Splitting proportion must be a value between 0 and 1.\n");
        exit(-1);
      }
    }

    // Read occurrences
    OccurrencesReader * occ = OccurrencesFactory::instance().create( source.c_str(), wkt.c_str() );

    OccurrencesPtr presences = occ->getPresences( label.c_str() );

    OccurrencesPtr absences = occ->getAbsences( label.c_str() );

    delete occ;

    // Split points if necessary
    if ( split_prop ) {

      OccurrencesPtr pres1, abs1, pres2, abs2;

      if ( presences ) {

        pres1 = new OccurrencesImpl( presences->label(), presences->coordSystem() );
        pres2 = new OccurrencesImpl( presences->label(), presences->coordSystem() );

        splitOccurrences( presences, pres1, pres2, split_prop );
      }

      if ( absences ) {

        abs1 = new OccurrencesImpl( absences->label(), absences->coordSystem() );
        abs2 = new OccurrencesImpl( absences->label(), absences->coordSystem() );

        splitOccurrences( absences, abs1, abs2, split_prop );
      }

      // Write result in two files

      ofstream outfile1( file1.c_str() );

      writeOutput( outfile1, format, pres1, abs1, label );

      outfile1.close();

      ofstream outfile2( file2.c_str() );

      writeOutput( outfile2, format, pres2, abs2, label );

      outfile2.close();

      return 0;
    }

    // Write result in cout

    std::cerr << flush;

    writeOutput( cout, format, presences, absences, label );
  }
  catch ( runtime_error e ) {

    printf( "om_points: %s\n", e.what() );
    exit(-1);
  }

  return 0;
}


// Function to output result
void writeOutput( ostream & stream, std::string format, OccurrencesPtr presences, OccurrencesPtr absences, std::string label ) {

    if ( format == "XML" ) {

      if ( presences ) {

        ConfigurationPtr cfg = presences->getConfiguration();
        cfg->setName( "Presence" );

        Configuration::writeXml( cfg, stream );
      }

      if ( absences && absences->numOccurrences() ) {

        ConfigurationPtr cfg = absences->getConfiguration();
        cfg->setName( "Absence" );

        Configuration::writeXml( cfg, stream );
      }
    }
    else {

      // Header
      stream << "#id\t" << "label\t" << "long\t" << "lat\t" << "abundance" << endl << flush;

      if ( presences ) {

        OccurrencesImpl::iterator it   = presences->begin();
        OccurrencesImpl::iterator last = presences->end();

        while ( it != last ) {

          stream << (*it)->id() << "\t" << label.c_str() << "\t" << (*it)->x() << "\t" << (*it)->y() << "\t" << (*it)->abundance() << endl << flush;
          ++it;
        }
      }

      if ( absences ) {

        OccurrencesImpl::iterator it   = absences->begin();
        OccurrencesImpl::iterator last = absences->end();

        while ( it != last ) {

          stream << (*it)->id() << "\t" << label.c_str() << "\t" << (*it)->x() << "\t" << (*it)->y() << "\t" << (*it)->abundance() << endl << flush;
          ++it;
        }
      }
    }
}
